File: backend/app/auth/branch_scope.py
from dataclasses import dataclass

from fastapi import HTTPException

from app.auth.permissions import SCOPES
from app.auth.schemas import CurrentUser


@dataclass
class BranchScope:
    ma_cn: str | None
    ma_pgd: str | None


def resolve_branch_scope(
    user: CurrentUser,
    requested_ma_cn: str | None = None,
    requested_ma_pgd: str | None = None,
) -> BranchScope:
    if user.can_view_all_branches():
        return BranchScope(ma_cn=requested_ma_cn, ma_pgd=requested_ma_pgd)

    home_cn = user.ma_cn
    home_pgd = user.ma_pgd
    allowed_branches = user.allowed_branches or ([home_cn] if home_cn else [])
    allowed_pgds = user.allowed_pgds or ([home_pgd] if home_pgd else [])

    if user.scope == SCOPES["PGD"]:
        if requested_ma_cn and requested_ma_cn != home_cn:
            raise HTTPException(status_code=403, detail="Không có quyền xem chi nhánh này")
        if requested_ma_pgd and requested_ma_pgd != home_pgd:
            raise HTTPException(status_code=403, detail="Không có quyền xem PGD này")
        return BranchScope(ma_cn=home_cn, ma_pgd=home_pgd)

    if requested_ma_cn and allowed_branches and requested_ma_cn not in allowed_branches:
        raise HTTPException(status_code=403, detail="Không có quyền xem chi nhánh này")

    ma_cn = requested_ma_cn or home_cn
    ma_pgd = requested_ma_pgd
    if ma_pgd and allowed_pgds and ma_pgd not in allowed_pgds:
        ma_pgd = None

    return BranchScope(ma_cn=ma_cn, ma_pgd=ma_pgd)

File: frontend/src/auth/branchScope.js
import { PERMISSIONS, SCOPES } from './permissions';

/**
 * Tính phạm vi chi nhánh từ user + lựa chọn UI.
 * Module quản trị chỉ cần đảm bảo /api/auth/me trả đúng shape user.
 */
export function resolveBranchScope(user, requestedCn = null, requestedPgd = null) {
  if (!user) {
    return {
      filterCn: requestedCn,
      filterPgd: requestedPgd,
      canViewProvince: true,
      canChangeBranch: true,
      canChangePgd: true,
      allowedBranches: [],
      allowedPgds: [],
      defaultCn: null,
      defaultPgd: null,
      denied: false,
    };
  }

  const permissions = user.permissions || [];
  const homeCn = user.ma_cn || user.branch_code || null;
  const homePgd = user.ma_pgd || user.department_code || null;
  const canViewAll =
    permissions.includes(PERMISSIONS.BRANCH_VIEW_ALL) ||
    permissions.includes(PERMISSIONS.ADMIN) ||
    user.scope === SCOPES.PROVINCE ||
    (!user.scope && !homeCn);
  const allowedBranches =
    user.allowed_branches?.length > 0
      ? user.allowed_branches
      : homeCn
        ? [homeCn]
        : [];

  const allowedPgds = user.allowed_pgds || (homePgd ? [homePgd] : []);

  if (canViewAll) {
    return {
      filterCn: requestedCn,
      filterPgd: requestedPgd,
      canViewProvince: true,
      canChangeBranch: true,
      canChangePgd: true,
      allowedBranches: [],
      allowedPgds: [],
      defaultCn: null,
      defaultPgd: null,
      denied: false,
    };
  }

  if (user.scope === SCOPES.PGD || user.scope === SCOPES.OWN) {
    const lockedCn = homeCn;
    const lockedPgd = homePgd;
    if (requestedCn && requestedCn !== lockedCn) {
      return deniedScope(lockedCn, lockedPgd, allowedBranches, allowedPgds);
    }
    if (requestedPgd && requestedPgd !== lockedPgd) {
      return deniedScope(lockedCn, lockedPgd, allowedBranches, allowedPgds);
    }
    return {
      filterCn: lockedCn,
      filterPgd: lockedPgd,
      canViewProvince: false,
      canChangeBranch: false,
      canChangePgd: false,
      allowedBranches,
      allowedPgds,
      defaultCn: lockedCn,
      defaultPgd: lockedPgd,
      denied: false,
    };
  }

  // scope branch (mặc định cán bộ chi nhánh)
  if (requestedCn && !allowedBranches.includes(requestedCn)) {
    return deniedScope(homeCn, null, allowedBranches, allowedPgds);
  }

  const filterCn = requestedCn ?? homeCn;
  let filterPgd = requestedPgd;
  if (filterPgd && allowedPgds.length > 0 && !allowedPgds.includes(filterPgd)) {
    filterPgd = null;
  }

  return {
    filterCn,
    filterPgd: filterPgd ?? null,
    canViewProvince: false,
    canChangeBranch: allowedBranches.length > 1,
    canChangePgd: allowedPgds.length === 0 || allowedPgds.length > 1,
    allowedBranches,
    allowedPgds,
    defaultCn: homeCn,
    defaultPgd: homePgd,
    denied: false,
  };
}

function deniedScope(cn, pgd, allowedBranches, allowedPgds) {
  return {
    filterCn: cn,
    filterPgd: pgd,
    canViewProvince: false,
    canChangeBranch: false,
    canChangePgd: false,
    allowedBranches,
    allowedPgds,
    defaultCn: cn,
    defaultPgd: pgd,
    denied: true,
  };
}

export function hasPermission(user, permission) {
  if (!user) return true;
  const permissions = user.permissions || [];
  return permissions.includes(PERMISSIONS.ADMIN) || permissions.includes(permission);
}

export function toApiBranchParams(filterCn, filterPgd) {
  const params = {};
  if (filterCn) params.ma_cn = filterCn;
  if (filterPgd) params.ma_pgd = filterPgd;
  return params;
}
